src/modules/fancyzones/lib/FancyZones.cpp

```diff
@@ -609,11 +610,20 @@ void FancyZones::OnDisplayChange(DisplayChangeType changeType) noexcept
             m_currentVirtualDesktopId = currentVirtualDesktopId;
             wil::unique_cotaskmem_string id;
             if (changeType == DisplayChangeType::Initialization &&
-                SUCCEEDED_LOG(StringFromCLSID(m_currentVirtualDesktopId, &id)))
+                SUCCEEDED(StringFromCLSID(m_currentVirtualDesktopId, &id)))
             {
                 JSONHelpers::FancyZonesDataInstance().UpdatePrimaryDesktopData(id.get());
             }
         }
+        if (changeType == DisplayChangeType::Initialization)
+        {
+            // Get currenty active desktops and clean data from persisted storage for all closed desktops.
+            std::vector<std::wstring> ids{};
+            if (VirtualDesktopUtils::GetVirtualDesktopIds(ids))
+            {
+                JSONHelpers::FancyZonesDataInstance().RemoveDeletedDesktops(ids);
+            }
+        }

```

Comment:
I found the code from line 603 to line 626 poorly readable.
First of all, the writelock at line 609 shouldn't be needed anymore.
The `if` at line 603-604 checks two `changeType(s)` then there are two more `if` for `changeType == DisplayChangeType::Initialization`, can we streamline the code to reduce the amounts of `if`?

Before refactoring the code, can we make `GetCurrentVirtualDesktopId` to always return the valid VD id? If it's the 0000 Guid let's get the real id and return it.
That should allow to refactor the code so that we don't do so many `if`.

Reply:
Nice catch for `if's`!

Reply:
I'll refactor the code to remove multiple checks for `changeType`, but `GetCurrentVirtualDesktopId` already work in the way You described.

Reply:
I moved `UpdatePrimaryDesktopData` to if statement below, as it only needs to be called on `DisplayChangeType::Initialization` and I am passing first element of array containing virtual desktop ids. First element is primary desktop, and it is exactly the same as passing `m_currentVirtualDesktopId` if we are in initialization phase. It reduces code complexity significantly here.